dali/imgcodec/parsers/png.cc

```diff
@@ -12,19 +12,99 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "dali/core/byte_io.h"
 #include "dali/imgcodec/parsers/png.h"
 
 namespace dali {
 namespace imgcodec {
 
+// https://www.w3.org/TR/2003/REC-PNG-20031110
+
+enum ColorType : uint8_t {
+  PNG_COLOR_TYPE_GRAY       = 0,
+  PNG_COLOR_TYPE_RGB        = 2,
+  PNG_COLOR_TYPE_PALETTE    = 3,
+  PNG_COLOR_TYPE_GRAY_ALPHA = 4,
+  PNG_COLOR_TYPE_RGBA       = 6
+};
+
+struct IhdrChunk {
+  uint32_t width;
+  uint32_t height;
+  uint8_t color_type;
+  // Some fields were ommited.
+
+  int GetNumberOfChannels() {
+    switch (color_type) {
+      case PNG_COLOR_TYPE_GRAY:
+      case PNG_COLOR_TYPE_GRAY_ALPHA:
+        return 1;
+      case PNG_COLOR_TYPE_RGB:
+      case PNG_COLOR_TYPE_PALETTE:  // 1 byte but it's converted to 3-channel BGR by OpenCV
+        return 3;
+      case PNG_COLOR_TYPE_RGBA:
+        return 4;
+      default:
+        DALI_FAIL(make_string("color type not supported: ", color_type));
+    }
+  }
+};
+
+using chunk_type_field_t = std::array<uint8_t, 4>;
+static constexpr chunk_type_field_t expected_ihdr_type_field = {73, 72, 68, 82};
+
+// Expects the read pointer in the stream to point to the beginning of a chunk.
+static IhdrChunk ReadIhdrChunk(InputStream& stream) {
+  // IHDR Chunk:
+  //  IHDR chunk length(4 bytes): 0x00 0x00 0x00 0x0D
+  //  IHDR chunk type(Identifies chunk type to be IHDR): 0x49 0x48 0x44 0x52
+  //  Image width in pixels(variable 4): xx xx xx xx
+  //  Image height in pixels(variable 4): xx xx xx xx
+  //  Flags in the chunk(variable 5 bytes): xx xx xx xx xx
+  //  CRC checksum(variable 4 bytes): xx xx xx xx
+
+  uint32_t length = ReadValueBE<uint32_t>(stream);
+  DALI_ENFORCE(length == 4 + 4 + 5);
+
+  chunk_type_field_t chunk_type;
+  stream.ReadAll<uint8_t>(chunk_type.data(), chunk_type.size());
+  DALI_ENFORCE(chunk_type == expected_ihdr_type_field);
```

Comment:
This is very hard to interpret compared to:
```suggestion
  char chunk_type[4];
  stream.ReadAll(chunk_type, 4);
  DALI_ENFORCE(!strncmp(chunk_type, "IHDR", 4));
```

Reply:
Why is your snippet less hard to interpret? Is it because of `"IHDR"` literal usage?

Reply:
Well, "constantitis" is just as bad as overuse of literals. A person vaguely familiar with PNG will recognize the string "IHDR" immediately. There are other ways to deal with it, if you have a beef against strncmp - there's the `tag` helper function used in `webp`, we can hoist it to some header. Then it would read:
```suggestion
  auto chunk_type =   stream.ReadOne<chunk_type>();
  DALI_ENFORCE(chunk_type == tag("IHDR"));
```

Reply:
> there's the tag helper function used in webp, we can hoist it to some header.

Which header seems appropriate to you?